File: src/main/java/com/zerozero/review/domain/service/ReadStoreReviewUseCase.java
package com.zerozero.review.domain.service;

import com.zerozero.review.domain.model.Review;
import com.zerozero.review.domain.response.ReviewResponse;
import com.zerozero.review.infrastructure.querydsl.ReviewQueryRepository;
import com.zerozero.store.presentation.request.ReadStoreRequest;
import com.zerozero.user.domain.model.User;
import com.zerozero.user.domain.repository.UserRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.log4j.Log4j2;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.Collections;
import java.util.List;
import java.util.Map;
import java.util.UUID;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
@Transactional(readOnly = true)
@Log4j2
public class ReadStoreReviewUseCase {

    private final ReviewQueryRepository reviewQueryRepository;

    private final UserRepository userRepository;

    public List<ReviewResponse> execute(ReadStoreRequest readStoreRequest, User user) {
        List<Review> reviews = reviewQueryRepository.findByStoreIdAndFilter(readStoreRequest.storeId(), readStoreRequest.filter());
        Map<UUID, User> reviewAuthorMap = getReviewAuthors(reviews);
        return convertReviewResponse(user, reviews, reviewAuthorMap);
    }

    private List<ReviewResponse> convertReviewResponse(User user, List<Review> reviews, Map<UUID, User> reviewAuthorMap) {
        if (user == null || reviews.isEmpty() || reviewAuthorMap.isEmpty()) {
            return Collections.emptyList();
        }
        return reviews.stream()
                .map(review -> {
                    User reviewAuthor = reviewAuthorMap.get(review.getUserId());
                    return ReviewResponse.of(review, reviewAuthor, review.getReviewLikes().size(), review.getReviewLikes().stream()
                            .anyMatch(like -> user.getId().equals(like.getUserId())));
                })
                .collect(Collectors.toList());
    }

    private Map<UUID, User> getReviewAuthors(List<Review> reviews) {
        return userRepository.findAllByIdInAndDeleted(
                reviews.stream()
                        .map(Review::getUserId)
                        .distinct()
                        .collect(Collectors.toList()),
                false
        ).stream().collect(Collectors.toMap(User::getId, reviewAuthor -> reviewAuthor));
    }

}

File: src/main/java/com/zerozero/review/domain/service/ReadStoreReviewUseCase.kt
package com.zerozero.review.domain.service

import com.zerozero.review.domain.model.Review
import com.zerozero.review.domain.response.ReviewResponse
import com.zerozero.review.infrastructure.querydsl.ReviewQueryRepository
import com.zerozero.store.presentation.request.ReadStoreRequest
import com.zerozero.user.domain.model.User
import com.zerozero.user.domain.repository.UserRepository
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.util.*

@Service
@Transactional(readOnly = true)
class ReadStoreReviewUseCase(
    private val reviewQueryRepository: ReviewQueryRepository,
    private val userRepository: UserRepository,
) {

    fun execute(readStoreRequest: ReadStoreRequest, user: User): List<ReviewResponse> {
        val reviews = reviewQueryRepository.findByStoreIdAndFilter(readStoreRequest.storeId, readStoreRequest.filter)
        val reviewAuthorMap = getReviewAuthors(reviews)
        return convertReviewResponse(user, reviews, reviewAuthorMap)
    }

    private fun convertReviewResponse(
        user: User,
        reviews: List<Review>,
        reviewAuthorMap: Map<UUID, User>,
    ): List<ReviewResponse> {
        if (reviews.isEmpty() || reviewAuthorMap.isEmpty()) return emptyList()

        return reviews.map { review ->
            val reviewAuthor = reviewAuthorMap[review.userId] ?: return@map null
            val isLiked = review.reviewLikes.any { it.userId == user.id }
            ReviewResponse.of(review, reviewAuthor, review.reviewLikes.size, isLiked)
        }.filterNotNull()
    }

    private fun getReviewAuthors(reviews: List<Review>): Map<UUID, User> {
        val userIds = reviews.map { it.userId }.distinct()
        val users = userRepository.findAllByIdIn(userIds)
        return users.associateBy { it.id }
    }
}
